### ISSUE TITLE
[Feature] Single-node end-to-end quickstart for agentic RL training

### ISSUE THREAD

**Issue by GJWu-zyx** [verl-project/uni-agent#114]
### Problem

New users who want to verify the full agentic RL training pipeline (rollout → reward → policy update) cannot do so on a single node. All existing training scripts require multi-node clusters (8-16 nodes), cloud sandbox platforms (vefaas/modal), and specific large models. The sandbox quickstart demo works great, but there is no equivalent for the training loop.

### Desired behavior

A single-node training quickstart that lets a user run the full agentic RL loop on 1 node with 1-8 GPUs/NPUs, using a local/Docker sandbox and a small publicly available model, with a minimal dummy dataset for smoke-testing.

### Owning component

Documentation or examples

### Target execution path

RL training

#### Clone and install
cd uni-agent && pip install -e .

#### Download a small model (e.g., Qwen2.5-7B)
huggingface-cli download Qwen/Qwen2.5-7B-Instruct --local-dir ./models/qwen2.5-7b

#### Run the single-node training quickstart (1 node, 8 GPUs)
bash examples/quickstart/training/train_single_node.sh

### Proposed design

A new shell script (e.g., train_single_node.sh) and task config that:
1. Sets NNODES=1, NGPUS_PER_NODE=8 (or lower), with appropriate parallelism settings
2. Uses sandbox.provider=local or docker instead of vefaas
3. Uses a small model path (user-provided or auto-downloaded)
4. Includes a minimal dummy dataset (a few samples) or a script to generate one
5. Disables optional dependencies (Wandb, cloud sandbox) by default
6. Logs to console instead of Wandb

The codebase already has all necessary abstractions (local sandbox, ReAct agent, task runner, framework adapter). Only configuration and documentation are missing.

### Compatibility and migration

No breaking changes. This is purely additive — a new example script and config.

### Alternatives considered

The only current alternative is to set up a multi-node cluster with cloud sandbox credentials, which is impractical for new users. The sandbox quickstart demo verifies the agent + tools stack but not the training loop.

### Validation

The quickstart script should:
1. Complete at least 2 training steps without errors
2. Show rollout metrics (success/failure counts)
3. Show reward values and policy update logs
4. Work on both CUDA and NPU single-node setups

### Additional context

- #43 is working on local Docker sandbox support for SWE-Bench
- examples/quickstart/sandbox/demo.py already verifies the agent + tools layer on a single node

### Contribution

- [x] I am willing to help implement this feature.
- [x] I am willing to help test or validate this feature.

**Comment by zackcxb** [comment]
Hi @GJWu-zyx, thank you for raising this issue. We agree that a lightweight, single-node end-to-end example covering the full rollout → reward → policy update pipeline would be valuable.
The recently added MemAgent recipe should already be runnable on a single node, and we are also working on a DeepEyes recipe that is single-node-runnable. That said, neither replaces a minimal, self-contained smoke test designed specifically to validate the full training pipeline with minimal setup.
Your proposed direction sounds reasonable and would be a valuable addition to the repository. We would be happy to review a PR for it. Thank you for offering to contribute!

**Comment by cyyecao-lappland** [comment]
Hi @GJWu-zyx — are you currently implementing this quickstart? I do not want to duplicate work you already have in progress.

If it is not underway, I would like to take the implementation. My proposed first deliverable is:

- one self-contained single-node script/config using a small public Qwen model;
- local or Docker sandbox, a tiny deterministic task dataset, console logging, and optional services such as W&B disabled by default;
- at least two completed rollout → reward → policy-update steps;
- explicit assertions/log evidence that rollout, reward, and optimizer update all occurred;
- documentation for 1/2/4-GPU scaling and common memory settings;
- CPU/config tests plus a real CUDA end-to-end validation on up to 4×RTX 3090 GPUs.

I do not have Ascend hardware, so I would not claim a real NPU execution result; I can keep the configuration backend-neutral and rely on project CI or an NPU maintainer for that part.

If you are already working on the implementation, I am also happy to coordinate instead and own the 4×3090 validation, regression checks, and documentation rather than competing with your PR.

**Comment by GJWu-zyx** [comment]
> Hi [@GJWu-zyx](https://github.com/GJWu-zyx) — are you currently implementing this quickstart? I do not want to duplicate work you already have in progress.
> 
> If it is not underway, I would like to take the implementation. My proposed first deliverable is:
> 
> * one self-contained single-node script/config using a small public Qwen model;
> * local or Docker sandbox, a tiny deterministic task dataset, console logging, and optional services such as W&B disabled by default;
> * at least two completed rollout → reward → policy-update steps;
> * explicit assertions/log evidence that rollout, reward, and optimizer update all occurred;
> * documentation for 1/2/4-GPU scaling and common memory settings;
> * CPU/config tests plus a real CUDA end-to-end validation on up to 4×RTX 3090 GPUs.
> 
> I do not have Ascend hardware, so I would not claim a real NPU execution result; I can keep the configuration backend-neutral and rely on project CI or an NPU maintainer for that part.
> 
> If you are already working on the implementation, I am also happy to coordinate instead and own the 4×3090 validation, regression checks, and documentation rather than competing with your PR.

Hi @cyyecao-lappland, thanks for reaching out and offering to help — your proposed deliverable looks really solid!

A quick update on the current state:

The MemAgent recipe ([examples/mem_agent/](vscode-webview://1bv45c553cgbirh8e52t4mfe916ekg80m7cg18q665ee3f5k2nf1/examples/mem_agent/)) is already in the repo and can run on a single node, though it's tied to the HotpotQA multi-hop QA task.
A DeepEyes adaptation is also in progress at #150.
Both of these are single-node runnable, but neither is quite the minimal, self-contained smoke test described in this issue — they are each tied to specific task domains, rather than being a lightweight pipeline check purely designed to verify that the full rollout → reward → policy update loop works end-to-end. So a truly minimal quickstart (small model + local/Docker sandbox + tiny dummy dataset + 2–3 steps to validate the full pipeline) would still be a valuable addition.

I'm not actively implementing it right now, so no duplication risk. If you'd like to take the lead, I'm happy to collaborate:

You own the GPU side — 4×RTX 3090 validation, regression checks, and documentation.
I'll handle the NPU side — validation on Ascend A2 hardware. We can share the same task config, and if the training scripts diverge enough, I'll write an NPU-specific version based on your GPU script, keeping the configuration backend-neutral.
Let me know if that works for you — happy to coordinate on the PR!

**Comment by cyyecao-lappland** [comment]
That works well for me — thanks!

I’ll take the lead on the GPU-side implementation and open an early draft PR so we can align on the shared task/config before any hardware-specific paths diverge.

I’ll keep the common configuration backend-neutral and focus the example on a minimal, deterministic pipeline smoke test, distinct from the task-specific MemAgent and DeepEyes recipes. The GPU validation will cover 2–3 complete rollout → reward → policy-update steps, regression checks, and the 1/2/4-GPU documentation, with the full run validated on 4×RTX 3090.

Once the shared GPU path is stable, I’ll tag you for Ascend A2 validation. If the NPU path needs a separate launcher, we can add that while continuing to share the same task configuration and expected validation signals.

I’ll link the draft PR here when the initial structure is ready.
